### PR TITLE
Shoot Quota support

### PR DESCRIPTION
**Pull request by dkistner**
Added shoot quota admission controller and quota control loop to shoot controller

See https://github.com/gardener/gardener/issues/81

### CODE REVIEW

File: pkg/apis/garden/types.go [hunk 1]
```diff
@@ -431,6 +438,28 @@ type QuotaSpec struct {
 	Scope QuotaScope
 }
 
+// QuotaStatus holds the most recently observed status of the Quota constraints.
+type QuotaStatus struct {
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: We do not have a `status` on `Quota` objects, so let's remove this.

File: pkg/controller/shoot/shoot.go [hunk 2]
```diff
@@ -160,20 +174,23 @@ func (c *Controller) Run(shootWorkers, shootCareWorkers, shootMaintenanceWorkers
 	}
 	for i := 0; i < shootMaintenanceWorkers; i++ {
 		controllerutils.CreateWorker(c.shootMaintenanceQueue, "Shoot Maintenance", c.reconcileShootMaintenanceKey, stopCh, &waitGroup, c.workerCh)
+		controllerutils.CreateWorker(c.shootQuotaQueue, "Shoot Quota", c.reconcileShootQuotaKey, stopCh, &waitGroup, c.workerCh)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot.go]: Can you please add configuration for this new controller (similar to [this](https://github.com/gardener/gardener/blob/master/pkg/apis/componentconfig/v1alpha1/types.go#L80-L84))?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	if err != nil {
+		return err
+	}
+
+	shoot, err := c.shootLister.Shoots(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Any reason why we don't have log messages here like in the [other controllers](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_maintenance_control.go#L62-L66)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	k8sGardenInformers gardeninformers.Interface
+}
+
+func (c *defaultQuotaControl) CheckQuota(shootObj *gardenv1beta1.Shoot, key string) error {
+	var (
+		clusterLifeTime     *int
+		quotaReferences     []v1.ObjectReference
+		shootExpirationTime time.Time
+
+		operationID = utils.GenerateRandomString(8)
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Operation IDs are only meant to be used for real cluster operations (reconcile, create, delete, maintenance). Please don't use them (see [this](https://github.com/gardener/gardener/blob/master/pkg/controller/shoot/shoot_care_control.go#L106)).

File: pkg/controller/shoot/shoot_quota_control.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = psb.Quotas
+	case "CrossSecretBinding":
+		csb, err := c.k8sGardenInformers.
+			CrossSecretBindings().
+			Lister().
+			CrossSecretBindings(shoot.Namespace).
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: This should be `shoot.Spec.Cloud.SecretBindingRef.Namespace`.

File: pkg/operation/common/types.go [hunk 6]
```diff
@@ -35,6 +35,11 @@ const (
 	// CloudPurposeSeed is a constant used while instantiating a cloud botanist for the Seed cluster.
 	CloudPurposeSeed = "seed"
 
+	// ClusterExpirationTime is an annotation on a Shoot resource whose value represent the time when the Shoot lifetime
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: *represents

File: pkg/operation/common/types.go [hunk 7]
```diff
@@ -35,6 +35,11 @@ const (
...
+	// is expired. The lifetime can be extended, but at most by the minimal value of the 'clusterLifetimeDays' property
+	// of referenced quotas.
+	ClusterExpirationTime = "expiration.garden.sapcloud.io/expirationTimestamp"
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: Hm, doubled `expiration`. Can we name that `shoot.garden.sapcloud.io/expirationTimestamp` (similar to [this](https://github.com/gardener/gardener/blob/master/pkg/operation/common/types.go#L155))?
And the constant should then have the name `ShootExpirationTimestamp` (by naming conventions for these labels/annotations).

File: plugin/pkg/global/resourcereferencemanager/admission.go [hunk 8]
```diff
@@ -172,10 +172,28 @@ func (r *ReferenceManager) ensureBindingReferences(secretNamespace, secretName s
...
+	)
+
 	for _, quotaRef := range quotaRefs {
-		if _, err := r.quotaLister.Quotas(quotaRef.Namespace).Get(quotaRef.Name); err != nil {
+		quota, err := r.quotaLister.Quotas(quotaRef.Namespace).Get(quotaRef.Name)
+		if err != nil {
 			return err
 		}
+		quotaScope := string(quota.Spec.Scope)
+		if quotaScope == quotaScopeProject {
```
Comment (rfranzke) [inline plugin/pkg/global/resourcereferencemanager/admission.go]: Why do  you need to convert this to `string`? `quotaScope` and `garden.QuotaScope[Secret|Project]` should be of the same type?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 9]
```diff
@@ -32,6 +39,24 @@ const (
 	PluginName = "ShootQuotaValidator"
 )
 
+var (
+	quotaMetricNames = [...]v1.ResourceName{
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: `[...]` is needed?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 10]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+
+	quotaReferences, err := h.getShootsQuotaReferences(*shoot)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	// Quotas are cumulative, means each quota must be not exceeded that the admission pass.
+	var maxShootLifetime *int
+	for _, quotaRef := range quotaReferences {
+		quota, err := h.quotaLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+			return err
+		}
+
+		// After the shoot has got a deletionTimestamp, we use this timestamp to set the
+		// deletionTimestampConfirmation annotation, which initiate the shoot deletion process.
+		annotations := shootUpdated.ObjectMeta.Annotations
+		annotations[common.ConfirmationDeletionTimestamp] = shootUpdated.DeletionTimestamp.Format(time.RFC3339)
+		shootUpdated.ObjectMeta.Annotations = annotations
+
+		_, err = c.k8sGardenClient.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+		// We have to delete the shoot, because only apiserver can set a deletionTimestamp
+		err := c.k8sGardenClient.
+			GardenClientset().
+			GardenV1beta1().
+			Shoots(shoot.Namespace).
+			Delete(shoot.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		shootUpdated, err := c.k8sGardenClient.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+			return err
+		}
+		shootExpirationTime = shoot.CreationTimestamp.Add(time.Duration(*clusterLifeTime*24) * time.Hour)
+	}
+
+	if time.Now().After(shootExpirationTime) {
+		shootLogger.Info("[SHOOT QUOTA] Shoot cluster lifetime expired. Shoot will be deleted.")
+
+		// We have to delete the shoot, because only apiserver can set a deletionTimestamp
+		err := c.k8sGardenClient.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+		if err != nil {
+			return err
+		}
+		shootExpirationTime = t.Add(time.Duration(*clusterLifeTime*24) * time.Hour)
+	} else {
+		annotations := shoot.Annotations
+		annotations[common.ClusterExpirationTime] = shoot.CreationTimestamp.Format(time.RFC3339)
+		shoot.Annotations = annotations
+
+		_, err := c.k8sGardenClient.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+		if err != nil {
+			return err
+		}
+		quotaReferences = csb.Quotas
+	default:
+		return fmt.Errorf("Unknown Binding type %s", shoot.Spec.Cloud.SecretBindingRef.Kind)
+	}
+
+	for _, quotaRef := range quotaReferences {
+		quota, err := c.k8sGardenInformers.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+			PrivateSecretBindings().
+			Lister().
+			PrivateSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = psb.Quotas
+	case "CrossSecretBinding":
+		csb, err := c.k8sGardenInformers.
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+
+		operationID = utils.GenerateRandomString(8)
+		shoot       = shootObj.DeepCopy()
+		shootLogger = logger.NewShootLogger(logger.Logger, shoot.Name, shoot.Namespace, operationID)
+	)
+
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "PrivateSecretBinding":
+		psb, err := c.k8sGardenInformers.
+			PrivateSecretBindings().
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 18]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
 	}
 
 	return nil
 }
+
+func (h *RejectShootIfQuotaExceeded) getShootsQuotaReferences(shoot garden.Shoot) ([]v1.ObjectReference, error) {
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "CrossSecretBinding":
+		usedCrossSB, err := h.crossSBLister.
+			CrossSecretBindings(shoot.Namespace).
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Should be `shoot.Spec.Cloud.SecretBindingRef.Namespace`.

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 19]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+		}
 	}
 
 	return nil
 }
+
+func (h *RejectShootIfQuotaExceeded) getShootsQuotaReferences(shoot garden.Shoot) ([]v1.ObjectReference, error) {
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "CrossSecretBinding":
+		usedCrossSB, err := h.crossSBLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 20]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+			CrossSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		return usedCrossSB.Quotas, nil
+	case "PrivateSecretBinding":
+		usedPrivateSB, err := h.privateSBLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 21]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+		}
+
+		exceededMetrics, err := h.isQuotaExceeded(*shoot, *quota)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+		if exceededMetrics != nil {
+			var message bytes.Buffer
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Any reason for a `Buffer` here?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 22]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+			}
+			if s.Spec.Cloud.SecretBindingRef.Kind == "PrivateSecretBinding" && s.Spec.Cloud.SecretBindingRef.Name == binding.Name {
+				shootsReferQuota = append(shootsReferQuota, *s)
+			}
+		}
+	}
+
+	// Find all shoots which are referencing found CrossSecretBindings
+	for _, binding := range crossSecretBindings {
+		shoots, err := h.shootLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 23]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+		return nil, err
+	}
+	crossSecretBindings, err := h.getCrossSecretBindingsReferQuota(quota)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find all shoots which are referencing found PrivateSecretBindings
+	for _, binding := range privateSecretBindings {
+		shoots, err := h.shootLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 24]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+			}
+		}
+	}
+	return shootsReferQuota, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) getPrivateSecretBindingsReferQuota(quota garden.Quota) ([]garden.PrivateSecretBinding, error) {
+	var privateSBsReferQuota []garden.PrivateSecretBinding
+
+	privateSBs, err := h.privateSBLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 25]
```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+			}
+		}
+	}
+	return privateSBsReferQuota, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) getCrossSecretBindingsReferQuota(quota garden.Quota) ([]garden.CrossSecretBinding, error) {
+	var crossSBsReferQuota []garden.CrossSecretBinding
+
+	crossSBs, err := h.crossSBLister.
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Can we write these statements into one line (line breakings are not used anywhere else in the code)?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/controller/shoot/shoot_quota_control.go (hunk 5 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: I thought the both kind of binding have to be in the same namespace as the shoot.
@mliepold mentioned that here in the story https://github.com/gardener/gardener/issues/81

The shoot resource dependency admission controller is also checking for bindings in the shoot namespace. https://github.com/gardener/gardener/blob/master/plugin/pkg/global/resourcereferencemanager/admission.go

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 18 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: See comment above to the shoot quota controller.

File: pkg/apis/garden/types.go (hunk 1 above)
Comment (dkistner) [reply pkg/apis/garden/types.go]: done.

File: pkg/controller/shoot/shoot.go (hunk 2 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot.go]: yes, added own configuration for the quota control loop.

File: pkg/controller/shoot/shoot_quota_control.go (hunk 3 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: Added additional log outputs here.

File: pkg/operation/common/types.go (hunk 6 above)
Comment (dkistner) [reply pkg/operation/common/types.go]: done

File: pkg/operation/common/types.go (hunk 7 above)
Comment (dkistner) [reply pkg/operation/common/types.go]: renamed it

File: plugin/pkg/global/resourcereferencemanager/admission.go (hunk 8 above)
Comment (dkistner) [reply plugin/pkg/global/resourcereferencemanager/admission.go]: yes it is the same type and conversion is not required. Refactored that.

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 9 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: Nope. Set this array to fixed size as the count of its elements is always fixed.

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 10 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 11 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 13 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 14 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 15 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 16 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 17 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: done

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 19 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 20 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 21 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: Buffer concatenation is more efficient (only O(n)) as concatenate strings directly. Think here we have at most six strings. Therefore switched to string concatenation.

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 23 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 24 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 25 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: done

File: pkg/controller/shoot/shoot_quota_control.go (hunk 5 above)
Comment (mliepold) [reply pkg/controller/shoot/shoot_quota_control.go]: Agree with @dkistner. In the current design with Private- and CrossSecrectBindings, there is no namespace in the secretBindingRef.
Only in the proposal from @mvladev that would remove the secret bindings there is a namespace in the secretBindingRef.

**Comment by rfranzke** [comment]
What's the status, can the PR be reviewed again? Can you rebase, please?

**Comment by dkistner** [comment]
Adapted the quota implementation to support only the new SecretBinding and not longer Cross/PrivateSecretBindings
Ensure that Shoot can be patched if its referenced quota is exceeded.

@rfranzke, @mliepold Could you please have a further look?

File: plugin/pkg/global/resourcereferencemanager/admission.go [hunk 26]
```diff
@@ -179,25 +179,25 @@ func (r *ReferenceManager) ensureSecretBindingReferences(attributes admission.At
...
-			APIVersion:      gardenv1beta1.SchemeGroupVersion.Version,
-			Resource:        "quotas",
-			Subresource:     "",
-			Namespace:       quotaRef.Namespace,
-			Name:            quotaRef.Name,
-			ResourceRequest: true,
-			Path:            "",
-		}
-		if decision, _, _ := r.authorizer.Authorize(readAttributes); decision != authorizer.DecisionAllow {
-			return errors.New("SecretBinding cannot reference a quota you are not allowed to read")
```
Comment (rfranzke) [inline plugin/pkg/global/resourcereferencemanager/admission.go]: Why has this authorization check been removed?

File: example/cloudprofile-openstack.yaml [hunk 27]
```diff
@@ -28,13 +28,13 @@ spec:
...
         memory: 4Gi
-        volumeType: default
+        volumeType: standard
         volumeSize: 20Gi
       - name: medium_4_8
         cpu: "4"
         gpu: "0"
         memory: 8Gi
-        volumeType: default
+        volumeType: standard
```
Comment (rfranzke) [inline example/cloudprofile-openstack.yaml]: Could you please rebase, change https://github.com/gardener/gardener/blob/master/hack/templates/resources/cloudprofile.yaml.tpl#L361 and https://github.com/gardener/gardener/blob/master/hack/templates/resources/cloudprofile.yaml.tpl#L367 and run `./hack/generate-examples`?

File: pkg/controller/shoot/shoot_quota_control.go [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+		annotations := shoot.Annotations
+		annotations[common.ShootExpirationTimestamp] = shoot.CreationTimestamp.Format(time.RFC3339)
+		shoot.Annotations = annotations
+
+		_, err := c.k8sGardenClient.GardenClientset().GardenV1beta1().Shoots(shoot.Namespace).Update(shoot)
+		if err != nil {
+			return err
+		}
+		shootExpirationTime = shoot.CreationTimestamp.Add(time.Duration(*clusterLifeTime*24) * time.Hour)
+	}
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_quota_control.go]: Not sure whether I understand the above section.
From the doc string of `ShootExpirationTimestamp`: "ShootExpirationTimestamp is an annotation on a Shoot resource whose value represents the time when the Shoot lifetime is expired."
Here, you are writing the Shoot's `.metadata.creationTimestamp` value into the annotation and compute the real expiration time afterwards. Why don't we write the value of `shoot.CreationTimestamp.Add(time.Duration(*clusterLifeTime*24) * time.Hour)` initially into the annotation?

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 29]
```diff
@@ -83,15 +117,485 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
 	shoot, ok := a.GetObject().(*garden.Shoot)
 	if !ok {
 		return apierrors.NewBadRequest("could not convert resource into Shoot object")
 	}
 
-	// TODO: implement quota limit check logic here
-	if false {
-		return admission.NewForbidden(a, fmt.Errorf("quota limits exceeded for %v", shoot.Name))
+	// Pass if the shoot is intended to get deleted
+	if shoot.DeletionTimestamp != nil && shoot.Annotations[common.ConfirmationDeletionTimestamp] != "" {
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: Let's just check the `.metadata.deletionTimestamp` to be not `nil` (the annotation is just a way to confirm, but the deletion time stamp is relevant).

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 30]
```diff
@@ -83,15 +117,485 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+	}
+	if oldLifetime != new.Annotations[common.ShootExpirationTimestamp] {
+		return true
+	}
+	return false
+}
+
+func quotaVerificationNeeded(new, old garden.Shoot, provider garden.CloudProvider) bool {
+	// Check for diff on addon nginx-ingress (addon requires to deploy a load balancer)
+	if old.Spec.Addons.NginxIngress.Enabled == false && new.Spec.Addons.NginxIngress.Enabled == true {
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: `.spec.addons` and `.spec.addons.nginx-ingress` could be `nil` which must be checked.

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 31]
```diff
@@ -83,15 +117,485 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+		if lifetime, exists := oldShoot.Annotations[common.ShootExpirationTimestamp]; exists {
+			oldExpirationTime, err = time.Parse(time.RFC3339, lifetime)
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+		} else {
+			oldExpirationTime = shoot.CreationTimestamp.Time
+		}
+
+		calulatedExpirationTime = oldExpirationTime.Add(time.Duration(*maxShootLifetime*24) * time.Hour)
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: See above comment.

File: plugin/pkg/shoot/quotavalidator/admission.go [hunk 32]
```diff
@@ -83,15 +117,485 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
...
+	return allocatedResources, nil
+}
+
+func (h *RejectShootIfQuotaExceeded) findShootsReferQuota(quota garden.Quota, shoot garden.Shoot) ([]garden.Shoot, error) {
+	var (
+		shootsReferQuota []garden.Shoot
+		secretBindings   []garden.SecretBinding
+	)
+
+	allSecrerBindings, err := h.secretBindingLister.SecretBindings(v1.NamespaceAll).List(labels.Everything())
```
Comment (rfranzke) [inline plugin/pkg/shoot/quotavalidator/admission.go]: typo: secrer -> secret

File: plugin/pkg/global/resourcereferencemanager/admission.go (hunk 26 above)
Comment (dkistner) [reply plugin/pkg/global/resourcereferencemanager/admission.go]: My fault, removed that accidentally during the rebase. Added this again. Sorry!

File: example/cloudprofile-openstack.yaml (hunk 27 above)
Comment (dkistner) [reply example/cloudprofile-openstack.yaml]: Thanks added this. 

Btw: We should also install a yaml python package e.g. pyyaml in ``./hack/generate-examples`` if not available.

File: pkg/controller/shoot/shoot_quota_control.go (hunk 28 above)
Comment (dkistner) [reply pkg/controller/shoot/shoot_quota_control.go]: Changed that behaviour. Now writing the calculated expiration time initially to the annotation.

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 29 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: yes, you are right. Checking for ``.metadata.deletionTimestamp`` is sufficient

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 30 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: yes right, added the missing checks

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 31 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: Fixed accordingly. Here we have the special case that the old version of the Shoot does not have a clusterDeletion timestamp annotated yet. Therefore we have to calculate the oldLifetime in this case.

File: plugin/pkg/shoot/quotavalidator/admission.go (hunk 32 above)
Comment (dkistner) [reply plugin/pkg/shoot/quotavalidator/admission.go]: fixed

**Comment by dkistner** [comment]
Thanks @rfranzke for the review. I have addressed your suggestions. Could you please have a further short look?

**Comment by rfranzke** [comment]
LGTM, thanks.

**Review by rfranzke — APPROVED** [review]

**Pull request merged**